docs/proposals/affinity.md

```diff
@@ -0,0 +1,341 @@
+S1 had a soft affinity rule "try to run me on nodes that are running pods from S2" then we would prefer
+to place a S1 pod that we are scheduling onto a node that is running a S2 pod, and also to place
+a S2 pod that we are scheduling onto a node that is running a S1 pod.
+
+As one further twist, although
+hard affinity is not symmetric, there is an implicit soft affinity rule corresponding to every
+hard affinity rule: if the pods of S1 have a hard affinity rule "run me on nodes that are running
+pods from S2" then it is not required that there be S1 pods on a node in order to schedule a S2 pod onto that node,
+but it would be better if there are.
+
+## Examples
+
+Here are some examples of how you would express various affinity and anti-affinity rules using the API we described.
+
+### Affinity
+
+In the examples below, the word "put" is intentionally ambiguous; the rules are the same
+whether "put" means "must put" (hard) or "try to put" (soft)--all that changes is whether
+the rule goes in HardAffinityExpressions or SoftAffinityExpressions. Also, some of the examples
+use "zone" and some use "node," just to make the examples more interesting; any of the examples
+with "zone" will also work for "node" if you change the `TopologyKey`, and vice-versa.
+
+* **Put the pod in zone Z**:
+Tricked you! It is not possible express this using the API described here. For this you should use a NodeConstraint.
+
+* **Put the pod in a zone that is running at least one pod from service S**:
+`{PodSelector: <selector that matches S's pods>, TopologyKey: "zone"}`
+
+* **Put the pod on a node that is already running a pod that requires a license for software package P**:
+Assuming pods that require a license for software package P have a label `{key=license, value=P}`:
+`{PodSelector: "license" In "P", TopologyKey: "node"}`
+
+* **Put this pod in the same zone as other pods from its same service**:
+Assuming pods from this pod's service have some label `{key=service, value=S}`:
+`{PodSelector: "service" In "S", TopologyKey: "zone"}`
+
+This last example illustrates a small issue with this API when it is used
+with a scheduler that processes the pending queue one pod at a time, like the current
+Kubernetes scheduler. The HardAffinityExpression
+`{PodSelector: "service" In "S", TopologyKey: "zone"}`
+only "works" once one pod from service S has been scheduled. But if all pods in service
+S have this HardAffinityExpression in their PodSpec, then the HardAffinityExpression will block the first
+pod of the service from ever scheduling, since it is only allowed to run in a zone with another pod from
+the same service. And of course that means none of the pods of the service will be able
+to schedule. This problem *only* applies to hard affinity, not soft affinity or hard or soft
+anti-affinity. There are at least two ways to solve this problem
+* short-term: when a controller creates pods from a PodTemplate, for exactly *one* of those
```

Comment:
That's a pretty good idea. The one corner case I can see is if you have two schedulers S1 and S2 that run in parallel and can both schedule any and all pods. Say there is a hard affinity selector that says all the pods from the collection must run on the same node. If S1 and S2 schedule the first two pods from the collection at the same time, with your rule they might put them on different nodes. This problem doesn't happen if you tweak the controller to remove the affinity selector from just one pod in the collection. But I guess we could say a runtime mechanism will detect this later and evict some pod(s) to restore the invariant. (As mentioned elsewhere in this doc, we need such a mechanism to ensure that affinity requirements continue to be met if a pod's labels are changed.)